lib/pure/ospaths.nim

```diff
@@ -157,64 +158,173 @@ const
     ## The character which separates the base filename from the extension;
     ## for example, the '.' in ``os.nim``.
 
+# TODO: MOVE to strutils, sequtils, or algorithms
+proc countWhile*(str: string, pred : proc(a:char):bool, prefix = true): int =
+  ## returns the number of elements in ``str`` that satisfy ``pred`` predicate.
+  ## without interruption, starting from beginning (when ``prefix`` is true)
+  ## or end (when ``prefix`` is false)
+  runnableExamples:
+    import sugar
+    doAssert countWhile("abc", a=>a == '/') == 0
+    doAssert countWhile("//abc", a=>a == '/') == 2
+    doAssert countWhile("//", a=>a == '/') == 2
+    doAssert countWhile("", a=>a == '/') == 0
+
+    doAssert countWhile("abc", a=>a == '/', prefix = false) == 0
+    doAssert countWhile("abc//", a=>a == '/', prefix = false) == 2
+    doAssert countWhile("//", a=>a == '/', prefix = false) == 2
+    doAssert countWhile("", a=>a == '/', prefix = false) == 0
+  let lenS = len(str)
+  if prefix:
+    for i in 0..<lenS:
+      if not pred(str[i]): return i
+  else:
+    for i in 0..<lenS:
+      if not pred(str[lenS - i - 1]): return i
+  return lenS
+
+# TODO: once slices are supported, define instead `rootPrefix`
+proc rootPrefixLength*(path: string) : int =
+  ## Returns the length of the prefix that makes ``path`` absolute, or 0
+  ## if ``path`` is relative.
+  runnableExamples:
+    doAssert "foo".rootPrefixLength == 0
+    doAssert "/foo".rootPrefixLength == 1
+    doAssert "//foo".rootPrefixLength == 2
+    when doslikeFileSystem:
+      doAssert r"\foo".rootPrefixLength == 1
+      doAssert r"/foo".rootPrefixLength == 1
+      doAssert r"C:".rootPrefixLength == 2
+      doAssert r"C:\foo".rootPrefixLength == 3
+      doAssert r"//foo".rootPrefixLength == 2
+      doAssert r"C:\\foo".rootPrefixLength == 4
+
+  if len(path) == 0: return 0
 
-proc joinPath*(head, tail: string): string {.
-  noSideEffect, rtl, extern: "nos$1".} =
-  ## Joins two directory names to one.
-  ##
-  ## For example on Unix:
-  ##
-  ## .. code-block:: nim
-  ##   joinPath("usr", "lib")
-  ##
-  ## results in:
-  ##
-  ## .. code-block:: nim
-  ##   "usr/lib"
+  when doslikeFileSystem:
+    if path[0] in {DirSep, AltSep}:
+      return countWhile(path, a => a in {DirSep, AltSep})
+    if len(path) > 1 and path[0] in {'a'..'z', 'A'..'Z'} and path[1] == ':':
+      # eg: C:\\bar
+      return 2 + countWhile(path[2..^1], a => a in {DirSep, AltSep})
+  elif defined(macos):
+    # according to https://perldoc.perl.org/File/Spec/Mac.html `:a` is a relative path
+    if path[0] == ':':
+      result = 0
+    else:
+      result = countWhile(path, a => a != ':')
```

Comment:
Is ``macos`` pre-mac-os-x only? If it covers latest macOS then this looks wrong to me.

Reply:
IIRC `macos` is there legacy classic Mac, `macosx` is, well, Mac OS X. This is kind of confusing though because since 10.12 Sierra they rebranded it as macOS.

Reply:
yeah, this is far too confusing.

Reply:
what is too confusing? the code in PR or `macos`? I already suggested removing support for `macos` https://forum.nim-lang.org/t/4039 "should we keep supporting macos? (as opposed to macosx)"